lib/engine/game/g_1849/step/buy_token.rb

```diff
@@ -0,0 +1,154 @@
+              @game.phase.tiles.include?(:green) &&
+              current_entity == entity &&
+              !@round.bought_token &&
+              !available_tokens(entity).empty? &&
+              MIN_PRICE <= buying_power(entity)
+          end
+
+          def available_tokens(entity)
+            entity.tokens_by_type
+          end
+
+          def can_sell_token?(token, entity)
+            token.corporation.placed_tokens.size > 1 &&
+            token != token.corporation.tokens.first &&
+            !token.city.tokened_by?(entity)
+          end
+
+          # look for any cities reachable from entity that are tokened by another corporation that
+          # has at least 2 tokens placed
+          #
+          # this is expensive - move to auto_actions
+          def any_buyable_tokens_placed?(entity)
+            @game.token_graph_for_entity(entity).connected_nodes(entity).keys.each do |node|
+              next unless node.city?
+
+              node.tokens.each do |token|
+                next unless token
+                next if token.corporation == entity
+
+                return true if can_sell_token?(token, entity)
+              end
+            end
+            false
+          end
+
+          def auto_actions(entity)
+            return [Engine::Action::Pass.new(entity)] unless any_buyable_tokens_placed?(entity)
+
+            super
+          end
+
+          # 1849 doesn't allow more than one token per corporation per tile
+          def can_token_city?(entity, city)
+            city.tile.nodes.select(&:city?).none? { |c| c.tokened_by?(entity) }
+          end
+
+          def can_replace_token?(entity, token)
+            return false unless token
+
+            other_corporation = token.corporation
+            city = token.city
+            entity != other_corporation &&
+              token.hex.id != other_corporation.tokens.first.hex.id &&
+              other_corporation.player &&
+              token.used &&
+              city &&
+              can_token_city?(entity, city) &&
+              (city.pass? || other_corporation.placed_tokens.count { |t| !t.city.pass? } > 1) &&
+              @game.token_graph_for_entity(entity).connected_nodes(entity)[city]
+          end
+
+          def description
+            'Buy a Token From Another Corporation'
+          end
+
+          def pass_description
+            'Skip (Buy a Token)'
+          end
+
+          def log_skip(entity)
+            return if !@game.acquiring_station_tokens? || !@game.phase.tiles.include?(:green)
+
+            @log << "#{entity.name} can't buy any tokens from other corporations"
+          end
+
+          def log_pass(entity)
+            @log << "#{entity.name} can't buy any tokens from other corporations"
+          end
+
+          def max_price(entity)
+            buying_power(entity)
+          end
+
+          def available_hex(entity, hex)
+            @game.token_graph_for_entity(entity).reachable_hexes(entity)[hex]
```

Comment:
Does `connected_nodes` align with `reachable_hexes`? I was thinking this would be `connected_hexes`.

Reply:
Okay, I know you've tested it, so if it works... Please verify you aren't able to buy a token in a non-connected hex that is upgradeable to make it connected.